File: internal/game/game.go
package game

import (
	"errors"
)

var (
	errTooSmallPlayers      = errors.New("game cannot be played with less than 2 players")
	errTooManyPlayers       = errors.New("game cannot be played with more than 4 players")
	errNegativePlayers      = errors.New("players cannot be less than 0")
	errNotPlayerTurn        = errors.New("not player's turn to move")
	errNilPointer           = errors.New("nil pointer error")
	errInvalidAttackingCell = errors.New("attacking cell is not owned by attacking player")
	errInvalidUpgradingCell = errors.New("upgrading cell is not owned by player")
	errGameAlreadyFinished  = errors.New("the game has already finished")
)

// Game represents the core structure that encapsulates the state and logic of the game.
// It holds references to the game board, a list of players, and the current turn's player ID.
type Game struct {
	Board      *Board   // Instance of the board
	Players    []Player // List of players
	turn       int      // ID of the player whose turn it is
	winnerId   int      // ID of the player who winned, -1 if the game is still on
	turnsLimit int      // Max count of turns in game
	turnsCount int      // Current count of turns
}

// createGame creates a new game with a given board and players.
func createGame(rows, cols, numPlayers int, seed int64, isFull bool) (*Game, error) {
	players, err := NewPlayersSlice(numPlayers)
	if err != nil {
		return nil, err
	}

	var board *Board
	if isFull {
		board, err = NewBoard(rows, cols)
	} else {
		board, err = NewRandomBoard(rows, cols, seed)
	}
	if err != nil {
		return nil, err
	}

	game, err := NewGameWithBoard(board, players)
	if err != nil {
		return nil, err
	}

	game.placePlayers()
	game.countPlayersCell()

	return game, nil
}

// NewGame creates a new Game with a random Board and specified number of players.
func NewGame(rows, cols int, numPlayers int, seed int64) (*Game, error) {
	return createGame(rows, cols, numPlayers, seed, false)
}

// NewCompleteBoardGame creates a new Game with a fully filled board (for testing purposes).
func NewCompleteBoardGame(rows, cols int, numPlayers int) (*Game, error) {
	return createGame(rows, cols, numPlayers, 0, true)
}

// NewGameWithBoard creates a new Game with a given Board and player list.
func NewGameWithBoard(board *Board, players []Player) (*Game, error) {
	if len(players) < 2 {
		return nil, errTooSmallPlayers
	} else if len(players) > 4 {
		return nil, errTooManyPlayers
	}

	game := &Game{
		Board:      board,
		Players:    players,
		turn:       0,
		winnerId:   -1,
		turnsLimit: board.cols * board.rows,
	}

	return game, nil
}

// NewPlayersSlice creates a slice of Player instances based on the given number.
func NewPlayersSlice(numPlayers int) ([]Player, error) {
	if numPlayers < 0 {
		return nil, errNegativePlayers
	}

	players := make([]Player, numPlayers)

	for i := range players {
		players[i] = newPlayer(i)
	}

	return players, nil
}

// Turn returns the ID of the current player's turn.
func (g *Game) Turn() int {
	return g.turn
}

func (g *Game) IsFinished() bool {
	return g.winnerId != -1
}

func (g *Game) Winner() Player {
	if g.IsFinished() {
		return g.Players[g.winnerId]
	} else {
		return nil
	}
}

// placePlayers places players on the board at specific locations.
func (g *Game) placePlayers() {
	const startPower = 2
	const startLevel = 1

	for idx, player := range g.Players {
		var c Cell
		switch idx {
		case 0:
			c = findNearestCell(g.Board, 0, 0)
		case 1:
			c = findNearestCell(g.Board, g.Board.rows-1, g.Board.cols-1)
		case 2:
			c = findNearestCell(g.Board, g.Board.rows-1, 0)
		case 3:
			c = findNearestCell(g.Board, 0, g.Board.cols-1)
		}
		g.Board.Cells[c.Row()][c.Col()] = newCellWithParameters(c.Row(), c.Col(), startLevel, startPower, player)
	}
}

// findNearestCell finds the nearest cell based on row and column coordinates.
func findNearestCell(board *Board, row, col int) Cell {
	// Offset for mirror columns
	var offset int
	if col > board.cols/2 {
		offset = 1
		if row%2 == 0 {
			offset = -offset
		}
	}

	var cell Cell
	for dist := 0; dist < max(board.cols, board.rows); dist++ {
		// Search left
		cell, _ = board.GetCell(Coords{row, col - dist})
		if cell != nil {
			return cell
		}
		// Search right
		cell, _ = board.GetCell(Coords{row, col + dist})
		if cell != nil {
			return cell
		}

		var locOffset int
		if dist%2 == 1 {
			locOffset = offset
		}

		// Search up
		cell, _ = board.GetCell(Coords{row - dist, col + locOffset})
		if cell != nil {
			return cell
		}
		// Search down
		cell, _ = board.GetCell(Coords{row + dist, col + locOffset})
		if cell != nil {
			return cell
		}
	}
	return nil
}

// countPlayersCell updates player cell counts on the board.
func (g *Game) countPlayersCell() {
	for _, row := range g.Board.Cells {
		for _, cell := range row {
			if cell != nil && cell.Owner() != nil {
				cell.Owner().addCell()
			}
		}
	}
}

// Attack performs an attack from one cell to another.
func (g *Game) Attack(player Player, from, to Cell) error {
	if g.IsFinished() {
		return errGameAlreadyFinished
	}
	if player.Id() != g.turn {
		return errNotPlayerTurn
	}
	if from == nil || to == nil {
		return errNilPointer
	}
	if from.Owner() != player {
		return errInvalidAttackingCell
	}
	if err := player.attack(); err != nil {
		return err
	}

	lastCellDestroyed, err := from.attack(to)
	if lastCellDestroyed {
		if lastPlayerId := g.findLastPlayerWithCells(); lastPlayerId != -1 {
			g.finish(lastPlayerId)
		}
	}

	return err
}

// EndAttack ends the attack phase for the current player.
func (g *Game) EndAttack(player Player) error {
	if g.IsFinished() {
		return errGameAlreadyFinished
	}
	if player.Id() != g.turn {
		return errNotPlayerTurn
	}

	return player.endAttack()
}

// Upgrade upgrades a target cell's level by a specified number of levels.
func (g *Game) Upgrade(player Player, target Cell, levels int) error {
	if g.IsFinished() {
		return errGameAlreadyFinished
	}
	if player.Id() != g.turn {
		return errNotPlayerTurn
	}
	if target == nil {
		return errNilPointer
	}
	if target.Owner() != player {
		return errInvalidUpgradingCell
	}
	if err := player.upgrade(target, levels); err != nil {
		return err
	}

	err := target.upgrade(levels)

	return err
}

// EndTurn ends the current player's turn, updating the board and switching turns.
func (g *Game) EndTurn(player Player) error {
	if g.IsFinished() {
		return errGameAlreadyFinished
	}
	if player.Id() != g.turn {
		return errNotPlayerTurn
	}
	player.endUpgrade()

	g.Board.calculatePower(player)

	g.nextTurn()

	return nil
}

// nextTurn advances the turn to the next player with CellsCount != 1.
func (g *Game) nextTurn() {
	currentPlayerIndex := g.turn
	nextPlayerIndex := (currentPlayerIndex + 1) % len(g.Players)
	foundPlayerIndex := -1

	// Iterate through the players to find the next suitable player.
	for nextPlayerIndex != currentPlayerIndex {
		if nextPlayerIndex == 0 {
			g.turnsCount++
			if g.turnsCount >= g.turnsLimit {
				g.finish(g.findPlayerWithMaxCells())
				return
			}
		}

		if g.Players[nextPlayerIndex].CellsCount() != 0 {
			foundPlayerIndex = nextPlayerIndex
			break
		}
		nextPlayerIndex = (nextPlayerIndex + 1) % len(g.Players)
	}

	if foundPlayerIndex != -1 {
		g.turn = foundPlayerIndex
	} else {
		// No other player found with non-zero CellsCount, finish the game.
		g.finish(g.turn)
	}
}

// findLastPlayerWithCells returns the id of the player with cells if he is alone, otherwise -1
func (g *Game) findLastPlayerWithCells() int {
	activePlayerCount := 0
	lastActivePlayerIndex := -1

	for i, player := range g.Players {
		if player.CellsCount() > 0 {
			activePlayerCount++
			lastActivePlayerIndex = i
		}
	}

	if activePlayerCount == 1 {
		return lastActivePlayerIndex
	}
	return -1
}

// findPlayerWithMaxCells returns the id of the player with the maximum number of cells
func (g *Game) findPlayerWithMaxCells() int {
	maxCells := 0
	playerId := -1

	for i, player := range g.Players {
		if player.CellsCount() >= maxCells {
			playerId = i
		}
	}

	return playerId
}

func (g *Game) finish(winnerId int) {
	g.winnerId = winnerId
}

func (g *Game) ToMap() map[string]interface{} {
	return map[string]interface{}{
		"board":     g.Board.toMap(),
		"players":   toPlayerInterfaceSlice(g.Players),
		"turn":      g.turn,
		"winner_id": g.winnerId,
	}
}

// ToMap converts the game state into a map for serialization.
func toPlayerInterfaceSlice(players []Player) []interface{} {
	result := make([]interface{}, len(players))
	for i, p := range players {
		result[i] = p.toMap()
	}
	return result
}

File: python/game.py
from random import randint


class Game:
    def reinit(self, game: dict) -> None:
        self.player = game['turn']
        self.game = game['board']['cells']
        self.points = game['players'][self.player]['points']
        self.actions = {'attack': None, 'upgrade': None}

    def doTurn(self) -> None:
        best_score, best_from, best_to = 0, (0, 0), (0, 0)
        for row in range(len(self.game)):
            for col in range(len(self.game[row])):
                if self.game[row][col] is None:
                    continue

                cell = (row, col)
                cell_power = self.get_cell(cell)['power']
                if self.owner(cell) == self.player and cell_power > 1:
                    for to in self.get_alien_neighbors(cell):
                        score = self.calculate_score(cell_power, to)
                        if score > best_score:
                            best_score = score
                            best_from = cell
                            best_to = to

        # print(f'FINAL: {best_score=}, {best_from=}, {best_to=}')
        if best_score:
            self.put_attack(best_from, best_to)

    def doUpgrade(self) -> None:
        if not self.points:
            return

        ownedCells = []
        for row in range(len(self.game)):
            for col in range(len(self.game[row])):
                if self.game[row][col] is None:
                    continue

                cell = (row, col)
                if self.owner(cell) == self.player:
                    ownedCells.append(cell)

        maxrow, maxcol = len(self.game), len(self.game[0])
        main_cell = {
            0: (0, 0),
            1: (maxrow, maxcol),
            2: (maxrow, 0),
            3: (0, maxcol)
        }[self.player]

        ownedCells.sort(key=lambda cell:
                        [abs(main_cell[0]-cell[0]) + abs(main_cell[1]-cell[1]),
                         -self.get_cell(cell)['level']],
                        reverse=True)

        for cell in ownedCells:
            cell_level = self.get_cell(cell)['level']
            upgradeCost = cell_level * (cell_level + 1) // 2
            if self.points >= upgradeCost:
                return self.put_upgrade(cell)

    def get_alien_neighbors(self, cell: tuple) -> None:
        offset = cell[0] % 2
        neighborsRelative = (
            (-1, offset - 1),  # up-left
            (-1, offset - 0),  # up-right
            (+1, offset - 1),  # down-left
            (+1, offset - 0),  # down-right
            (0, -1),           # left
            (0, +1),           # right
        )

        for row, col in neighborsRelative:
            to = (cell[0] + row, cell[1] + col)
            if (self.get_cell(to) is not None
               and self.owner(to) != self.player):
                yield to

    def calculate_score(self, from_power, to) -> int:
        # If cell is free
        if self.owner(to) is None:
            return from_power

        # If our cell power lower than other
        to_power = self.get_cell(to)['power']
        if from_power <= to_power:
            return 100 + (to_power - from_power)

        # Otherwise
        return 10000 + (to_power - from_power)

    def put_attack(self, cell: tuple, to: tuple) -> None:
        self.actions['attack'] = [cell, to]

    def put_upgrade(self, cell: tuple) -> None:
        self.actions['upgrade'] = list(cell)

    def get_cell(self, cell: tuple) -> dict:
        if (cell[0] < 0 or cell[1] < 0 or cell[0] >= len(self.game)
           or cell[1] >= len(self.game[cell[0]])):
            return None
        return self.game[cell[0]][cell[1]]

    def owner(self, cell: tuple) -> int:
        return self.get_cell(cell).get('owner_id', None)
